File: test/File.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeOperators #-}

module File where
-- base
import Data.Proxy
import GHC.Generics (Generic)
-- aeson
import Data.Aeson
-- QuickCheck
import Test.QuickCheck
-- quickcheck-arbitrary-adt
import Test.QuickCheck.Arbitrary.ADT
-- ocaml-export
import OCaml.Export

type FilePackage = OCamlPackage "" NoDependency :>
  (OCamlModule '["File"]
    :> OCamlTypeInFile Person "test/ocaml/Person"
    :> Automobile
    :> OCamlTypeInFile Business "test/ocaml/Business"
    :> OCamlTypeInFile (Wrapper TypeParameterRef0 TypeParameterRef1) "test/ocaml/Wrapper"
    :> AutoDependingOnManual
    :> HaskellTypeName "NonGenericType" (OCamlTypeInFile NonGenericType "test/ocaml/NonGenericType")
  )

data Person = Person
  { id :: Int
  , name :: Maybe String
  } deriving (Show, Eq, Generic, FromJSON, ToJSON)

instance Arbitrary Person where
  arbitrary = Person <$> arbitrary <*> arbitrary

instance ToADTArbitrary Person

instance OCamlType Person where
  toOCamlType _ = typeableToOCamlType (Proxy :: Proxy Person)

data Automobile = Automobile
  { make :: String
  , model :: String
  , year :: Int
  } deriving (Show, Eq, Generic, OCamlType, FromJSON, ToJSON)

instance Arbitrary Automobile where
  arbitrary = Automobile <$> arbitrary <*> arbitrary <*> arbitrary

instance ToADTArbitrary Automobile

data Business = Business
  { taxId :: String
  , owner :: Person
  , employees :: [Person]
  , companyVehicle :: Maybe Automobile
  } deriving (Show, Eq, Generic, FromJSON, ToJSON)

instance Arbitrary Business where
  arbitrary = do
    employeeCount <- choose (0,3)
    Business <$> arbitrary <*> arbitrary <*> vector employeeCount <*> arbitrary

instance ToADTArbitrary Business

instance OCamlType Business where
  toOCamlType _ = typeableToOCamlType (Proxy :: Proxy Business)

data Wrapper a b = Wrapper
  { wrapperA :: a
  , wrapperB :: b
  , wrapperC :: String
  } deriving (Show, Eq, Generic, FromJSON, ToJSON)

instance (Arbitrary a, Arbitrary b) => Arbitrary (Wrapper a b) where
  arbitrary = Wrapper <$> arbitrary <*> arbitrary <*> arbitrary

instance (Arbitrary a, ToADTArbitrary a, Arbitrary b, ToADTArbitrary b) => ToADTArbitrary (Wrapper a b)

instance OCamlType (Wrapper TypeParameterRef0 TypeParameterRef1) where
  toOCamlType _ = typeableToOCamlType (Proxy :: Proxy (Wrapper TypeParameterRef0 TypeParameterRef1))

data AutoDependingOnManual = AutoDependingOnManual
  { abc :: String
  , bbBusiness :: Business
  } deriving (Show, Eq, Generic, OCamlType, FromJSON, ToJSON)

instance Arbitrary AutoDependingOnManual where
  arbitrary = AutoDependingOnManual <$> arbitrary <*> arbitrary

instance ToADTArbitrary AutoDependingOnManual

data NonGenericType =
  NonGenericType
    { ngA :: String
    , ngB :: Int
    } deriving (Show, Eq)

instance ToJSON NonGenericType where
  toJSON ng = object [ "ngA" .= ngA ng, "ngB" .= ngB ng]

instance FromJSON NonGenericType where
  parseJSON = withObject "NonGenericType" $ \o ->
    NonGenericType <$> o .: "ngA" <*> o .: "ngB"

instance ToADTArbitrary NonGenericType where
  toADTArbitrarySingleton Proxy =
    ADTArbitrarySingleton "File" "NonGenericType"
      <$> (ConstructorArbitraryPair "NonGenericType" <$> (NonGenericType <$> arbitrary <*> arbitrary))

  toADTArbitrary Proxy =
    ADTArbitrary "File" "NonGenericType"
      <$> sequence
        [ ConstructorArbitraryPair "NonGenericType" <$> (NonGenericType <$> arbitrary <*> arbitrary)
        ]

instance OCamlType NonGenericType where
  toOCamlType _ = typeableToOCamlType (Proxy :: Proxy NonGenericType)

File: ocaml-export/test/interface/golden/file/File.ml
type person =
  { id : int
  ; name : (string) option
  }

let encodePerson x =
  Aeson.Encode.object_
    [ ( "id", Aeson.Encode.int x.id )
    ; ( "name", Aeson.Encode.optional Aeson.Encode.string x.name )
    ]

let decodePerson json =
  match Aeson.Decode.
    { id = field "id" int json
    ; name = optional (field "name" string) json
    }
  with
  | v -> Belt.Result.Ok v
  | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("decodePerson: " ^ message)


type automobile =
  { make : string
  ; model : string
  ; year : int
  }

let encodeAutomobile x =
  Aeson.Encode.object_
    [ ( "make", Aeson.Encode.string x.make )
    ; ( "model", Aeson.Encode.string x.model )
    ; ( "year", Aeson.Encode.int x.year )
    ]

let decodeAutomobile json =
  match Aeson.Decode.
    { make = field "make" string json
    ; model = field "model" string json
    ; year = field "year" int json
    }
  with
  | v -> Belt.Result.Ok v
  | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("decodeAutomobile: " ^ message)

type business =
  { taxId : string
  ; owner : person
  ; employees : (person) list
  ; companyVehicle : (automobile) option
  }

let encodeBusiness x =
  Aeson.Encode.object_
    [ ( "taxId", Aeson.Encode.string x.taxId )
    ; ( "owner", encodePerson x.owner )
    ; ( "employees", (Aeson.Encode.list encodePerson) x.employees )
    ; ( "companyVehicle", Aeson.Encode.optional encodeAutomobile x.companyVehicle )
    ]

let decodeBusiness json =
  match Aeson.Decode.
    { taxId = field "taxId" string json
    ; owner = field "owner" (fun a -> unwrapResult (decodePerson a)) json
    ; employees = field "employees" (list (fun a -> unwrapResult (decodePerson a))) json
    ; companyVehicle = optional (field "companyVehicle" (fun a -> unwrapResult (decodeAutomobile a))) json
    }
  with
  | v -> Belt.Result.Ok v
  | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("decodeBusiness: " ^ message)


type ('a, 'b) wrapper =
  { wrapperA : 'a
  ; wrapperB : 'b
  ; wrapperC : string
  }

let encodeWrapper a b x =
  Aeson.Encode.object_
    [ ( "wrapperA", a x.wrapperA )
    ; ( "wrapperB", b x.wrapperB )
    ; ( "wrapperC", Aeson.Encode.string x.wrapperC )
    ]

let decodeWrapper a b json =
  match Aeson.Decode.
        { wrapperA = field "wrapperA" (fun x -> unwrapResult (a x)) json
        ; wrapperB = field "wrapperB" (fun x -> unwrapResult (b x)) json
        ; wrapperC = field "wrapperC" string json
        }
  with
  | v -> Belt.Result.Ok v
  | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("decodeWrapper: " ^ message)


type autoDependingOnManual =
  { abc : string
  ; bbBusiness : business
  }

let encodeAutoDependingOnManual x =
  Aeson.Encode.object_
    [ ( "abc", Aeson.Encode.string x.abc )
    ; ( "bbBusiness", encodeBusiness x.bbBusiness )
    ]

let decodeAutoDependingOnManual json =
  match Aeson.Decode.
    { abc = field "abc" string json
    ; bbBusiness = field "bbBusiness" (fun a -> unwrapResult (decodeBusiness a)) json
    }
  with
  | v -> Belt.Result.Ok v
  | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("decodeAutoDependingOnManual: " ^ message)

type nonGenericType =
  { ngA : string
  ; ngB : int
  }

let encodeNonGenericType x =
  Aeson.Encode.object_
    [ ( "ngA", Aeson.Encode.string x.ngA)
    ; ( "ngB", Aeson.Encode.int x.ngB)
    ]

let decodeNonGenericType json =
  match Aeson.Decode.
        { ngA = field "ngA" string json
        ; ngB = field "ngB" int json
        }
  with
  | v -> Belt.Result.Ok v
  | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("decodeNonGenericType: " ^ message)

